### PR TITLE
abci: application evidence prepared by evidence pool

### PR DESCRIPTION
**Pull request by cmwaters**
## Description

Implementation of sending evidence from the evidence pool to the application as outlined in [ADR59](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-059-evidence-composition-and-lifecycle.md)

### CODE REVIEW

File: p2p/mocks/peer.go [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// Code generated by mockery v2.2.1. DO NOT EDIT.
+// Code generated by mockery v2.1.0. DO NOT EDIT.
```
Comment (cmwaters) [inline p2p/mocks/peer.go:1]: hmm I wonder if I have a different version of mockery installed

File: evidence/pool.go [hunk 2]
```diff
@@ -207,6 +209,82 @@ func (evpool *Pool) CheckEvidence(evList types.EvidenceList) error {
...
+		key := keyCommitted(ev)
+
+		h := gogotypes.Int64Value{Value: height}
+		evBytes, err := proto.Marshal(&h)
+		if err != nil {
+			panic(err)
+		}
+
+		if err := evpool.evidenceStore.Set(key, evBytes); err != nil {
+			evpool.logger.Error("Unable to add committed evidence", "err", err)
```
Comment (melekes) [inline evidence/pool.go:230]: continue?

File: evidence/verify.go [hunk 3]
```diff
@@ -233,25 +246,34 @@ func getMaliciousValidators(evidence *types.LightClientAttackEvidence, commonVal
...
+		!bytes.Equal(trusted.NextValidatorsHash, conflicting.NextValidatorsHash) ||
+		!bytes.Equal(trusted.ConsensusHash, conflicting.ConsensusHash) ||
+		!bytes.Equal(trusted.AppHash, conflicting.AppHash) ||
+		!bytes.Equal(trusted.LastResultsHash, conflicting.LastResultsHash)
 }
+
+type lightClientAttackType int
+
+const (
+	lunaticType lightClientAttackType = iota
```
Comment (melekes) [inline evidence/verify.go:276]:
```suggestion
	lunaticType lightClientAttackType = iota+1
```

**Review by melekes — APPROVED** [review]
🚀

File: evidence/pool.go (hunk 2 above)
Comment (cmwaters) [reply evidence/pool.go:230]: I will shift it to the bottom. Adding the evidence to the committed list should be the last thing we do

**Pull request merged**